contracts/EscrowAndStaking/StakingThales.sol

```diff
@@ -599,12 +684,118 @@ contract StakingThales is IStakingThales, Initializable, ProxyOwned, ProxyReentr
         }
         stakerAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = stakerAMMVolume[account][
             periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)
-        ]
-            .amount
-            .add(amount);
+        ].amount.add(amount);
+
+        if (msg.sender == thalesAMM) {
+            if (lastThalesAMMUpdatePeriod[account] < periodsOfStaking) {
+                thalesAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = 0;
+                thalesAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].period = periodsOfStaking;
+                lastThalesAMMUpdatePeriod[account] = periodsOfStaking;
+            }
+            thalesAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = thalesAMMVolume[account][
+                periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)
+            ].amount.add(amount);
+        }
+
+        if (msg.sender == thalesRangedAMM) {
+            if (lastThalesRangedAMMUpdatePeriod[account] < periodsOfStaking) {
+                thalesRangedAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = 0;
+                thalesRangedAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].period = periodsOfStaking;
+                lastThalesRangedAMMUpdatePeriod[account] = periodsOfStaking;
+            }
+            thalesRangedAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = thalesRangedAMMVolume[
+                account
+            ][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount.add(amount);
+        }
+
+        if (msg.sender == exoticBonds) {
+            if (lastExoticMarketsUpdatePeriod[account] < periodsOfStaking) {
+                exoticMarketsVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = 0;
+                exoticMarketsVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].period = periodsOfStaking;
+                lastExoticMarketsUpdatePeriod[account] = periodsOfStaking;
+            }
+            exoticMarketsVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = exoticMarketsVolume[
+                account
+            ][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount.add(amount);
+        }
+
+        if (msg.sender == sportsAMM) {
+            if (lastSportsAMMUpdatePeriod[account] < periodsOfStaking) {
+                sportsAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = 0;
+                sportsAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].period = periodsOfStaking;
+                lastSportsAMMUpdatePeriod[account] = periodsOfStaking;
+            }
+            sportsAMMVolume[account][periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)].amount = sportsAMMVolume[account][
+                periodsOfStaking.mod(AMM_EXTRA_REWARD_PERIODS)
+            ].amount.add(amount);
```

Comment:
it might make sense to trigger event for each different AMM, 
or (better) extend the event with a flag of which AMM updated the volume.

Reply:
The event is not used by subgraph or UI, because we cannot track volume for the last 4 epochs via events. So, no need to extend more.